test/src/test/java/org/corfudb/runtime/view/LogUnitSealTest.java

```diff
@@ -0,0 +1,105 @@
+                .buildStripe()
+                .addLogUnit(SERVERS.PORT_0)
+                .addToSegment()
+                .addToLayout()
+                .build();
+
+        bootstrapAllServers(l);
+        CorfuRuntime corfuRuntime = getRuntime(l).connect();
+        RuntimeLayout runtimeLayout = corfuRuntime.getLayoutView().getRuntimeLayout(l);
+
+        getManagementServer(SERVERS.PORT_0).shutdown();
+
+        return runtimeLayout;
+    }
+
+    /**
+     * Test that log unit server seals correctly.
+     * Check operations in the queue that come after the seal operation, but stamped with
+     * an older epoch, are completed exceptionally with a WrongEpochException.
+     */
+    @Test
+    public void checkOperationWithOldEpochFailedAfterSeal() {
+        long epoch = 1L;
+        RuntimeLayout runtimeLayout = getRuntimeLayout(epoch);
+        Layout layout = runtimeLayout.getLayout();
+        LogUnitClient client = runtimeLayout.getLogUnitClient(SERVERS.ENDPOINT_0);
+
+        List<CompletableFuture<Boolean>> successOpFutures1 = new ArrayList<>();
+        List<CompletableFuture<Boolean>> successOpFutures2 = new ArrayList<>();
+        List<CompletableFuture<Boolean>> failedOpFutures = new ArrayList<>();
+
+        final int numOp = 3;
+        long address = 0L;
+
+        for (int i = 0; i < numOp; i++) {
+            LogData ld = LogData.getEmpty(address++);
+            successOpFutures1.add(client.write(ld));
+        }
+
+        // Seal server
+        layout.setEpoch(layout.getEpoch() + 1);
+        runtimeLayout.sealMinServerSet();
+
+        // Intentionally reset router epoch to bypass epoch check on router.
+        getLayoutServer(SERVERS.PORT_0).getServerContext().getServerRouter().setServerEpoch(epoch);
+
+        // Still using the old client to send messages stamped with old epoch
+        for (int i = 0; i < numOp; i++) {
+            LogData ld = LogData.getEmpty(address++);
+            failedOpFutures.add(client.write(ld));
+        }
+
+        // Set the router epoch to new epoch and get the new client
+        getLayoutServer(SERVERS.PORT_0).getServerContext().getServerRouter().setServerEpoch(layout.getEpoch());
+        runtimeLayout = new RuntimeLayout(layout, getRuntime(layout));
+        client = runtimeLayout.getLogUnitClient(SERVERS.ENDPOINT_0);
+
+        // Using the new client to send messages stamped with epoch 2
+        for (int i = 0; i < numOp; i++) {
+            LogData ld = LogData.getEmpty(address++);
+            successOpFutures2.add(client.write(ld));
+        }
+
+        // Before sealing, all operations should complete normally
+        successOpFutures1.forEach(f -> assertThatCode(f::join).doesNotThrowAnyException());
+        // After sealing, operations stamped with old epoch should fail
+        failedOpFutures.forEach(f -> assertThatThrownBy(f::join)
+                .isInstanceOf(CompletionException.class)
+                .hasCauseExactlyInstanceOf(WrongEpochException.class));
```

Comment:
How do you know that this exception is thrown by BatchWriter and not the base server?

Reply:
I manually reset the router's epoch: https://github.com/CorfuDB/CorfuDB/blob/5af64b5457606f4f6bf46c18812871945cb84bc4/test/src/test/java/org/corfudb/runtime/view/LogUnitSealTest.java#L77 

Also I verify it through the test logs.